packages/ogimage-gen/src/OgImageMiddleware.ts

```diff
@@ -0,0 +1,344 @@
+    width: number
+    height: number
+  }
+  format: {
+    [key in SUPPORTED_EXT]: {
+      type: string
+      quality?: number
+    }
+  }
+}
+
+interface ComponentElementProps {
+  Component: React.FC<{ data: unknown }>
+  data: unknown
+  routeParams: Record<string, unknown>
+  debug: boolean
+}
+
+export default class OgImageMiddleware {
+  options: MwOptions
+  App: React.FC
+  Document: React.FC<{ css: string[]; meta: string[] }>
+
+  // Initialized in invoke() 👇
+  imageProps?: {
+    width: number
+    height: number
+    quality: number
+  }
+
+  constructor(options: MwOptions) {
+    this.options = options
+
+    this.App = options.App
+    this.Document = options.Document
+  }
+
+  async invoke(
+    req: MiddlewareRequest,
+    mwResponse: MiddlewareResponse,
+    invokeOptions: MiddlewareInvokeOptions,
+  ) {
+    const url = new URL(req.url)
+    const { pathname } = url
+
+    let currentRoute: RWRouteManifestItem | undefined = undefined
+    let parsedParams: {
+      params?: Record<string, unknown>
+    } = {}
+
+    // Skip processing if not a file request
+    if (!pathname.includes('.')) {
+      return mwResponse
+    }
+
+    // Remove the extension for the match
+    const [routePathname, extension] = pathname.split('.')
+
+    ;({ currentRoute, parsedParams } = await this.matchRoute(
+      // This is a special case for the index route
+      // because they can't go to mywebsite.com/.png -> mywebsite.com/index.png instead
+      routePathname === '/index' ? '/' : routePathname,
+      parsedParams,
+    ))
+
+    // If no match with the router, or not a supported extension bail
+    if (!currentRoute || !supportedExtensions.includes(extension)) {
+      console.log(
+        'OGMiddleware: No match with the Routes, or not a supported extension',
+      )
+      return mwResponse
+    }
+
+    // Combine search params and params from route pattern
+    // /user/{id:Int} => /user/1?background=red => { id: 1, background: 'red'}
+    const mergedParams = {
+      ...Object.fromEntries(url.searchParams.entries()),
+      ...(parsedParams.params || {}),
+    }
+
+    this.imageProps = {
+      width: mergedParams.width
+        ? parseInt(mergedParams.width as string)
+        : OGIMAGE_DEFAULTS.width,
+      height: mergedParams.height
+        ? parseInt(mergedParams.height as string)
+        : OGIMAGE_DEFAULTS.height,
+      quality: mergedParams.quality
+        ? parseInt(mergedParams.quality as string)
+        : OGIMAGE_DEFAULTS.quality,
+    }
+
+    const debug = !!mergedParams.debug
+
+    const screenshotOptions: ScreenshotOptions = {
+      viewport: {
+        width: this.imageProps.width,
+        height: this.imageProps.height,
+      },
+      format: {
+        png: { type: 'png' },
+        jpg: {
+          type: 'jpeg',
+          quality: this.imageProps.quality,
+        },
+      },
+    }
+
+    // @TODO (ROB!)
+    // I think it doesn't work with jsx paths in my project IN DEV at the moment
```

Comment:
Can confirm! It looks like this is because the `relativeFilePath` is the path to the page itself, and that's still a .tsx file (I only renamed the og:image component to `.jsx`). So it looks for the same named file but with `.png.tsx` on the end, which doesn't exist. If you rename the page itself to also be `.jsx` then it works! So they both need to be `.jsx` or both `.tsx`.

Personally I'm okay with this limitation, I feel like most people are going to consistently have one or the other. If we wanted to be nice we could try/catch that error and say "Your og:image component needs to have the same extension as its parent page!", otherwise we'd have to do some file system shenanigans to try and be smart and convert one to the other, ugh.

Reply:
Will remove the TODO